src/de/dkfz/b080/co/knowledge/metadata/COMetadataAccessor.groovy

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (c) 2018 German Cancer Research Center (DKFZ).
+ *
+ * Distributed under the MIT License (license terms are at https://www.github.com/eilslabs/Roddy/LICENSE.txt).
+ */
+package de.dkfz.b080.co.knowledge.metadata
+
+import de.dkfz.b080.co.common.BasicCOProjectsRuntimeService
+import de.dkfz.b080.co.common.COConfig
+import de.dkfz.b080.co.common.MetadataTable
+import de.dkfz.b080.co.common.RunID
+import de.dkfz.b080.co.files.BasicBamFile
+import de.dkfz.b080.co.files.COFileStageSettings
+import de.dkfz.b080.co.files.Sample
+import de.dkfz.roddy.Roddy
+import de.dkfz.roddy.StringConstants
+import de.dkfz.roddy.core.DataSet
+import de.dkfz.roddy.core.ExecutionContext
+import de.dkfz.roddy.core.ExecutionContextError
+import de.dkfz.roddy.execution.io.MetadataTableFactory
+import de.dkfz.roddy.execution.io.fs.FileSystemAccessProvider
+import de.dkfz.roddy.knowledge.files.BaseFile
+import de.dkfz.roddy.tools.LoggerWrapper
+import de.dkfz.roddy.tools.RoddyIOHelperMethods
+import de.dkfz.roddy.tools.Tuple2
+import groovy.transform.CompileDynamic
+import groovy.transform.CompileStatic
+import sun.reflect.generics.reflectiveObjects.NotImplementedException
+
+import static de.dkfz.b080.co.files.COConstants.*
+
+@CompileStatic
+class COMetadataAccessor {
+
+    private static LoggerWrapper logger = LoggerWrapper.getLogger(COMetadataAccessor.class.getName())
+
+    private final static List<File> alreadySearchedMergedBamFolders = []
+
+    BasicCOProjectsRuntimeService runtimeService
+
+    COMetadataAccessor(BasicCOProjectsRuntimeService runtimeService) {
+        this.runtimeService = runtimeService
+    }
+
+    MetadataTable getMetadataTable(ExecutionContext context) {
+        return new MetadataTable(MetadataTableFactory.getTable(context.getAnalysis()).subsetByDataset(context.getDataSet().id))
+    }
+
+    protected List<Sample> extractSamplesFromMetadataTable(ExecutionContext context) {
+        return getMetadataTable(context).listColumn(INPUT_TABLE_SAMPLECOL_NAME).unique().collect {
+            new Sample(context, it)
+        }
+    }
+
+    protected List<Sample> extractSamplesFromFastqList(List<File> fastqFiles, ExecutionContext context) {
+        COConfig cfg = new COConfig(context)
+        return grepPathElementFromFilenames(cfg.getSequenceDirectory(), '${sample}', fastqFiles).unique().collect { Tuple2<String, File> pair ->
+            // TODO: Do we want a test here, that two files with the same sample (i.e. both matching the pattern!) come from the same sample directory?
```

Comment:
The question is not whether this is a weird file organisation. Unless it is logically possible to happen (because the code admits), the question is rather what is the risk to the user if he weirdly messes up his filestructure and how likely is it to happen!